diff --git a/ndk_cc_toolchain_config.bzl b/ndk_cc_toolchain_config.bzl
index e5796b5..2475919 100644
--- a/ndk_cc_toolchain_config.bzl
+++ b/ndk_cc_toolchain_config.bzl
@@ -289,6 +289,7 @@ def ndk_cc_toolchain_config(
         # Blaze requests this feature by default.
         # Blaze tests if this feature is supported before setting includes.
         feature(name = "include_paths"),
+        feature(name = "external_include_paths"),
 
         # Blaze tests if this feature is enabled in order to create implicit
         # "nodeps" .so outputs from cc_library rules.
@@ -900,6 +901,17 @@ def ndk_cc_toolchain_config(
                         ),
                     ],
                 ),
+                flag_set(
+                    actions = actions.preprocessor_compile,
+                    flag_groups = [
+                        flag_group(
+                            flags = ["-isystem", "%{external_include_paths}"],
+                            iterate_over = "external_include_paths",
+                            expand_if_available = "external_include_paths",
+                        ),
+                    ],
+                    features = ["external_include_paths"],
+                ),
                 flag_set(
                     actions = actions.preprocessor_compile,
                     flag_groups = [
